src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -1398,5 +1381,183 @@ private ImmutableArray<T> InsertSpanRangeInternal(int index, ReadOnlySpan<T> ite
+
+                _bucketHeaders = new int[capacity];
+                _entries = new Entry[capacity];
+            }
+
+            public MultiSet(IEqualityComparer<T>? equalityComparer)
+            {
+                _equalityComparer = equalityComparer ?? EqualityComparer<T>.Default;
+
+                _bucketHeaders = new int[1];
+                _entries = new Entry[1];
+            }
+
+            public void Add(T item)
+            {
+                ref int bucketHeader = ref GetBucketHeaderIndexRef(item);
+                int entryIndex = bucketHeader - 1;
+                while (entryIndex != -1)
+                {
+                    ref Entry entry = ref _entries[entryIndex];
+                    if (_equalityComparer.Equals(entry.Key, item))
+                    {
+                        entry.Count++;
+                        return;
+                    }
+
+                    entryIndex = entry.Next;
+                }
+
+                if (_startOffset == _bucketHeaders.Length)
+                {
+                    EnsureCapacity();
+                    bucketHeader = ref GetBucketHeaderIndexRef(item);
+                }
+
+                _entries[_startOffset].SetInfo(item, bucketHeader - 1, 1);
+                bucketHeader = ++_startOffset; // 1-based bucketHeader
+            }
+
+            private void EnsureCapacity()
+            {
+                const uint arrayMaxLength = 0X7FFFFFC7;
+
+                int expectedSize = _bucketHeaders.Length + 1;
+                uint newSize = (uint)_bucketHeaders.Length << 1;
+                if (newSize > arrayMaxLength)
+                {
+                    newSize = arrayMaxLength;
+                }
+
+                if (newSize < expectedSize)
+                {
+                    newSize = (uint)expectedSize;
+                }
+
+                var entries = new Entry[newSize];
+                Array.Copy(_entries, entries, _entries.Length);
+                _bucketHeaders = new int[newSize];
+                for (int i = 0; i < _entries.Length; i++)
+                {
+                    ref Entry entry = ref _entries[i];
+                    if (entry.Count != 0)
+                    {
+                        ref int bucketHeader = ref GetBucketHeaderIndexRef(entry.Key);
+                        entry.Next = bucketHeader - 1;
+                        bucketHeader = i + 1;
+                    }
+                }
+
+                _entries = entries;
+            }
+
+            public bool TryRemove(T item)
+            {
+                ref int bucketHeader = ref GetBucketHeaderIndexRef(item);
+                int entryIndex = bucketHeader - 1;
+                while (entryIndex != -1)
+                {
+                    ref Entry entry = ref _entries[entryIndex];
+                    if (_equalityComparer.Equals(entry.Key, item))
+                    {
+                        if (entry.Count == 0)
```

Comment:
Nit: not important in this particular application, but a general-purpose MultiSet should probably also remove the entry when count drops to zero. Consider adding a remark stating that we're not doing this.

Reply:
Added.